### PR TITLE
docs: add CDK examples

### LINKED ISSUE #104: Document how to use this with an `AwsCdkConstructLibrary` projen project
A common use-case for this package is using it with an [AWS CDK Construct Library](https://projen.io/awscdk-construct.html) with Projen. However, it's not immediately clear how to use it. The README shows `new ProjenStruct`, but it doesn't say where to write that code.

### PR DESCRIPTION
**Pull request by blimmer**
Fixes #104

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,16 +41,16 @@ Complex types can be used and will be imported using their FQN.
 Any existing properties of the same name will be replaced.
 
 ```ts
-new ProjenStruct(project, { name: 'MyProjectOptions'})
+new ProjenStruct(project, { name: 'MyProjectOptions' })
```
Comment (blimmer) [inline README.md:44]: `prettier` made some changes - lmk if you want me to back them out

File: README.md [hunk 2]
```diff
@@ -17,13 +17,19 @@ Install with:
 npm install --save-dev @mrgrain/jsii-struct-builder
 ```
 
+Then, place a `new ProjenStruct` in your `.projenrc.js` file, passing your
+[TypeScript project](https://projen.io/typescript.html) as the first parameter. See the sections below for more usage
+details.
+
+If you're not using [`projen`](https://projen.io/), see [Use without `projen`](#use-without-projen), below.
```
Comment (mrgrain) [inline README.md:24]:
```suggestion
Or if you're not using [`projen`](https://projen.io/), see [Use without `projen`](#use-without-projen).
```

File: README.md [hunk 3]
```diff
@@ -17,13 +17,19 @@ Install with:
 npm install --save-dev @mrgrain/jsii-struct-builder
 ```
 
+Then, place a `new ProjenStruct` in your `.projenrc.js` file, passing your
+[TypeScript project](https://projen.io/typescript.html) as the first parameter. See the sections below for more usage
+details.
```
Comment (mrgrain) [inline README.md:22]:
```suggestion
Then add a `new ProjenStruct` in your `.projenrc.ts` file, passing a [TypeScript project](https://projen.io/typescript.html) as the first parameter.
See the sections below for more usage details.
```

File: README.md [hunk 4]
```diff
@@ -104,18 +110,70 @@ new ProjenStruct(project, { name: 'MyProjectOptions'})
...
+});
+
+new ProjenStruct(project, { name: 'MyFunctionProps' })
+  .mixin(Struct.fromFqn('aws-cdk-lib.aws_lambda.FunctionProps'))
+  .withoutDeprecated()
+  .allOptional()
+  .omit('code'); // our construct always provides the code
+```
+
+Then, expose the new struct in your CDK construct.
```
Comment (mrgrain) [inline README.md:135]:
```suggestion
### AWS CDK properties

A common use-case of this project is to expose arbitrary overrides in CDK constructs.
For example, you may want to provide common AWS Lambda configuration, but allow a consuming user to override any arbitrary property.

To accomplish this, first create the new struct in your `.projenrc.ts` file.

```ts
import { ProjenStruct, Struct } from '@mrgrain/jsii-struct-builder';
import { awscdk } from 'projen';

declare project: awscdk.AwsCdkConstructLibrary;

new ProjenStruct(project, { name: 'MyFunctionProps' })
  .mixin(Struct.fromFqn('aws-cdk-lib.aws_lambda.FunctionProps'))
  .withoutDeprecated()
  .allOptional()
  .omit('code'); // our construct always provides the code
```

Then, expose the new struct in your CDK construct:
```

File: README.md [hunk 5]
```diff
@@ -104,18 +110,70 @@ new ProjenStruct(project, { name: 'MyProjectOptions'})
...
+
+const project = new awscdk.AwsCdkConstructLibrary({
+  // your config - see https://projen.io/awscdk-construct.html
+});
+
+new ProjenStruct(project, { name: 'MyFunctionProps' })
+  .mixin(Struct.fromFqn('aws-cdk-lib.aws_lambda.FunctionProps'))
+  .withoutDeprecated()
+  .allOptional()
+  .omit('code'); // our construct always provides the code
```
Comment (mrgrain) [inline README.md:132]:
```
```ts
import { ProjenStruct, Struct } from '@mrgrain/jsii-struct-builder';
import { awscdk } from 'projen';

declare project: awscdk.AwsCdkConstructLibrary;

new ProjenStruct(project, { name: 'MyFunctionProps' })
  .mixin(Struct.fromFqn('aws-cdk-lib.aws_lambda.FunctionProps'))
  .withoutDeprecated()
  .allOptional()
  .omit('code'); // our construct always provides the code
```

File: README.md [hunk 6]
```diff
@@ -104,18 +110,70 @@ new ProjenStruct(project, { name: 'MyProjectOptions'})
   .withoutDeprecated();
 ```
 
+### Exposing CDK Properties
+
+A common use-case for this project is to expose arbitrary overrides in CDK constructs. For example, you may want to
+provide common AWS Lambda configuration, but allow a consuming user to override any arbitrary property.
+
+To accomplish this, first create the new struct in your `.projenrc.js` file.
```
Comment (mrgrain) [inline README.md:118]:
```suggestion
### AWS CDK properties

A common use-case of this project is to expose arbitrary overrides in CDK constructs.
For example, you may want to provide common AWS Lambda configuration, but allow a consuming user to override any arbitrary property.

To accomplish this, first create the new struct in your `.projenrc.ts` file.
```

File: README.md (hunk 4 above)
Comment (mrgrain) [inline README.md:135]:
```suggestion
Then use the new struct in your CDK construct.
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
Thanks for this PR so far!  I made some minor suggestions, feel free to accept them or ignore them for now as I can always add them in before the merge.

One question remains though: I'm not entirely sure I follow the CDK use-case example.
You use `ProjenStruct` to create a new `MyFunctionProps` - but then in `lib/MyFunction.ts` you create that again and also use `FunctionOverrides`.

If the brief is to have everything from `FunctionProps` except `code`, then we should not need the extra steps via `FunctionOverrides`. Is that what your intention was here?

File: README.md (hunk 5 above)
Comment (blimmer) [reply README.md:132]: See comment above, RE: ts projen config file

**Comment by blimmer** [comment]
Many CDK constructs have properties that you can only configure at instantiation time. Therefore, if you're the author of a CDK construct and want to provide escape hatches for users, this is the pattern you'd use.

In my example, for instance, say the user wants to override the `description` on the function. Instead of explicitly exposing a custom variable declaration for just that property, I accept _any_ other `FunctionProp` and just arbitrarily let the user override it.

**Comment by mrgrain** [comment]
> Many CDK constructs have properties that you can only configure at instantiation time. Therefore, if you're the author of a CDK construct and want to provide escape hatches for users, this is the pattern you'd use.
> 
> In my example, for instance, say the user wants to override the `description` on the function. Instead of explicitly exposing a custom variable declaration for just that property, I accept _any_ other `FunctionProp` and just arbitrarily let the user override it.

Let me rephrase my question: The example is referencing a `FunctionOverrides` type but it's not clear where this is coming from or what it is.

File: README.md (hunk 3 above)
Comment (mrgrain) [reply README.md:22]: Yes!

https://projen.io/typescript.html#migrating-projenrcjs-to-projenrcts

New projects are setup with it by default! Looks like your projen version is out of date 😬 
```console
$ npx projen@latest new awscdk-construct
Need to install the following packages:
  projen@0.71.91
Ok to proceed? (y) y
👾 Project definition file was created at /Users/kornherm/code/scratchpad/pts/.projenrc.ts
```

File: README.md [hunk 7]
```diff
@@ -104,18 +109,70 @@ new ProjenStruct(project, { name: 'MyProjectOptions'})
...
 ```
 
+### AWS CDK properties
+
+A common use-case of this project is to expose arbitrary overrides in CDK constructs.
+For example, you may want to provide common AWS Lambda configuration, but allow a consuming user to override any arbitrary property.
+
+To accomplish this, first create the new struct in your `.projenrc.ts` file.
+
+```js
```
Comment (mrgrain) [inline README.md:119]:
```suggestion
```ts
```

**Comment by mrgrain** [comment]
> Ah, yep, I had a naming mismatch! Thanks. I think I've addressed all points of feedback.

You have, thank you!

Now allow me the question why we need to do the extra steps with `FunctionOverrides`.

Wouldn't this be a nicer pattern?

```ts
// .projenrc.ts
import { ProjenStruct, Struct } from '@mrgrain/jsii-struct-builder';
import { awscdk } from 'projen';
const project = new awscdk.AwsCdkConstructLibrary({
  // your config - see https://projen.io/awscdk-construct.html
});
new ProjenStruct(project, { name: 'MyFunctionProps' })
  .mixin(Struct.fromFqn('aws-cdk-lib.aws_lambda.FunctionProps'))
  .withoutDeprecated()
  .allOptional()
  .omit('code'); // our construct always provides `code`
```

```ts
// lib/MyFunction.ts
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { join } from 'path';
import { MyFunctionProps } from './MyFunctionProps';

export class MyFunction extends Construct {
  constructor(scope: Construct, id: string, props: MyFunctionProps = {}) {
    super(scope, id);
    const { functionOverrides = {} } = props;
    new Function(this, 'Function', {
      // sensible defaults
      runtime: Runtime.NODEJS_18_X,
      handler: 'index.handler',
      // user provided props
      ...props,
      // always force `code` to the provided one
      code: Code.fromAsset(join(__dirname, 'lambda-handler')),
    });
  }
}
```

**Comment by blimmer** [comment]
I guess I'm going off the pattern I use in a more complex construct. Surely there will be other configurations the user wants to provide in their standard interface, which I was trying to convey in my example. e.g., https://github.com/blimmer/cdk-static-wordpress#escape-hatches

However, I can simplify it down to any suggestion you prefer. With any additional documentation proposed here, I would have been able to more easily figure out what I implemented in cdk-static-wordpress 😄

**Review by mrgrain — APPROVED** [review]

**Comment by mrgrain** [comment]
Thanks again for improving the docs!
